src/mpid/ch4/shm/xpmem/xpmem_control.h

```diff
@@ -0,0 +1,154 @@
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef XPMEM_CONTROL_H_INCLUDED
+#define XPMEM_CONTROL_H_INCLUDED
+
+#include "ch4_impl.h"
+#include "xpmem_pre.h"
+#include "xpmem_impl.h"
+#include "xpmem_recv.h"
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_ctrl_send_lmt_ack_cb
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_ctrl_send_lmt_ack_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *sreq = (MPIR_Request *) ctrl_hdr->xpmem_slmt_ack.sreq_ptr;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_CTRL_SEND_LMT_ACK_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_CTRL_SEND_LMT_ACK_CB);
+
+    XPMEM_PT2PT_DBG_PRINT("send_lmt_ack_cb: complete sreq %p\n", sreq);
+    MPID_Request_complete(sreq);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_CTRL_SEND_LMT_ACK_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+#undef FUNCNAME
+#define FUNCNAME MPIDI_XPMEM_ctrl_send_lmt_req_cb
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_ctrl_send_lmt_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_req_t *slmt_req_hdr = &ctrl_hdr->xpmem_slmt_req;
+    MPIR_Request *rreq = NULL;
+    MPIR_Comm *root_comm;
+    MPIR_Request *anysource_partner;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_CTRL_SEND_LMT_REQ_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_CTRL_SEND_LMT_REQ_CB);
+
+    XPMEM_PT2PT_DBG_PRINT
+        ("send_lmt_req_cb: received src_offset 0x%lx, data_sz 0x%lx, sreq_ptr 0x%lx, "
+         "src_lrank %d, match info[src_rank %d, tag %d, context_id 0x%x]\n",
+         slmt_req_hdr->src_offset, slmt_req_hdr->data_sz, slmt_req_hdr->sreq_ptr,
+         slmt_req_hdr->src_lrank, slmt_req_hdr->src_rank, slmt_req_hdr->tag,
+         slmt_req_hdr->context_id);
+
+    /* Try to match a posted receive request. */
+    root_comm = MPIDIG_context_id_to_comm(slmt_req_hdr->context_id);
+    if (root_comm) {
```

Comment:
What are the situations where this would be `NULL`? If a message arrives before the communicator has been created or after it's been freed? Wouldn't that be an error?

Reply:
This code is basically copied from `MPIDIG_send_target_msg_cb`. I think when there is a matching posted receive request, `root_comm == NULL` can never happen because its recount cannot be zero. Thus, we need to handle such a case when posting an unexpected request (See code below). Actually `MPIDIG_enqueue_unexp` always enqueues request into a global list in current code.

Let me add a short comment here. I just spent about 10 minutes to understand the logic :(
```
/* Enqueue unexpected receive request */
if (rreq == NULL) {
...
        if (root_comm) {
            MPIR_Comm_add_ref(root_comm);       /* +1 for unexp_list */
            MPIDIG_enqueue_unexp(rreq, &MPIDIG_COMM(root_comm, unexp_list));
        } else {
            MPIDIG_enqueue_unexp(rreq,
                                 MPIDIG_context_id_to_uelist(MPIDIG_REQUEST(rreq, context_id)));
        }
}
```